group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilder.java

```diff
@@ -0,0 +1,327 @@
+            memberSpecs,
+            member,
+            assignments.getOrDefault(memberId, Assignment.EMPTY)
+        ));
+
+        // Update the member spec if updated or deleted members.
+        updatedMembers.forEach((memberId, updatedMemberOrNull) -> {
+            if (updatedMemberOrNull == null) {
+                memberSpecs.remove(memberId);
+            } else {
+                addMemberSpec(
+                    memberSpecs,
+                    updatedMemberOrNull,
+                    assignments.getOrDefault(memberId, Assignment.EMPTY)
+                );
+            }
+        });
+
+        // Prepare the topic metadata.
+        Map<Uuid, AssignmentTopicMetadata> topics = new HashMap<>();
+        subscriptionMetadata.forEach((topicName, topicMetadata) ->
+            topics.put(topicMetadata.id(), new AssignmentTopicMetadata(topicMetadata.numPartitions()))
+        );
+
+        // Compute the assignment.
+        GroupAssignment newGroupAssignment = assignor.assign(new AssignmentSpec(
+            Collections.unmodifiableMap(memberSpecs),
+            Collections.unmodifiableMap(topics)
+        ));
+
+        // Compute delta from previous to new assignment and create the
+        // relevant records.
+        List<Record> records = new ArrayList<>();
+        Map<String, Assignment> newTargetAssignment = new HashMap<>();
+
+        memberSpecs.keySet().forEach(memberId -> {
+            Assignment oldMemberAssignment = assignments.get(memberId);
+            Assignment newMemberAssignment = newMemberAssignment(newGroupAssignment, memberId);
+
+            newTargetAssignment.put(memberId, newMemberAssignment);
+
+            if (oldMemberAssignment == null) {
+                // If the member had no assignment, we always create a record for it.
+                records.add(newTargetAssignmentRecord(
+                    groupId,
+                    memberId,
+                    newMemberAssignment.partitions()
+                ));
+            } else {
+                // If the member had an assignment, we only create a record if the
+                // new assignment is different.
+                if (!newMemberAssignment.equals(oldMemberAssignment)) {
+                    records.add(newTargetAssignmentRecord(
+                        groupId,
+                        memberId,
+                        newMemberAssignment.partitions()
+                    ));
+                }
+            }
+        });
+
+        // Bump the assignment epoch.
+        records.add(newTargetAssignmentEpochRecord(groupId, groupEpoch));
+
+        return new TargetAssignmentResult(records, newTargetAssignment);
+    }
+
+    private Assignment newMemberAssignment(
+        GroupAssignment newGroupAssignment,
+        String memberId
+    ) {
+        MemberAssignment newMemberAssignment = newGroupAssignment.members().get(memberId);
+        if (newMemberAssignment != null) {
```

Comment:
when is the assignment null? When we are removing the group?

Reply:
this is a defensive measure to protect us from the case where the assignor would not return an assignment for a member.

Reply:
Ok -- so we don't typically expect this, but it isn't bad enough to throw an error.

Reply:
Yeah, it is much better to handle this case.